src/mongocxx/test/spec/unified_tests/operations.cpp

```diff
@@ -553,6 +553,112 @@ document::value fail_point(entity::map& map, spec::apm_checker& apm, document::v
+    document::view filter = arguments["filter"].get_document().value;
+    options::find_one_and_delete options{};
+
+    if (arguments["collation"]) {
+        options.collation(arguments["collation"].get_document().value);
+    }
+    if (arguments["hint"]) {
+        if (arguments["hint"].type() == bsoncxx::v_noabi::type::k_string)
+            options.hint(hint{arguments["hint"].get_string().value});
+        else
+            options.hint(hint{arguments["hint"].get_document().value});
+    }
+    if (arguments["projection"]) {
+        options.projection(arguments["projection"].get_document().value);
+    }
+    if (arguments["sort"]) {
+        options.sort(arguments["sort"].get_document().value);
+    }
+
+    stdx::optional<document::value> document;
+    if (session) {
+        document = coll.find_one_and_delete(*session, filter, options);
+    } else {
+        document = coll.find_one_and_delete(filter, options);
+    }
+
+    // Server versions below 3.0 sometimes return an empty document rather than null when no
+    // documents match.
+    auto result = builder::basic::document{};
+    if (document && !(document->view().empty())) {
+        result.append(builder::basic::kvp("result", *document));
+    } else {
+        result.append(builder::basic::kvp("result", types::b_null{}));
+    }
+
+    return result.extract();
+}
+
+document::value find_one_and_replace(collection& coll,
+                                     client_session* session,
+                                     document::view operation) {
+    document::view arguments = operation["arguments"].get_document().value;
+    document::view filter = arguments["filter"].get_document().value;
+    options::find_one_and_replace options{};
+
+    if (arguments["collation"]) {
+        options.collation(arguments["collation"].get_document().value);
+    }
+    if (arguments["hint"]) {
+        if (arguments["hint"].type() == bsoncxx::v_noabi::type::k_string)
+            options.hint(hint{arguments["hint"].get_string().value});
+        else
+            options.hint(hint{arguments["hint"].get_document().value});
+    }
+    if (arguments["projection"]) {
+        options.projection(arguments["projection"].get_document().value);
+    }
+    if (arguments["returnDocument"]) {
+        auto return_document = string::to_string(arguments["returnDocument"].get_string().value);
+
+        if (return_document == "After") {
+            options.return_document(options::return_document::k_after);
+        }
+
+        if (return_document == "Before") {
+            options.return_document(options::return_document::k_before);
+        }
+    }
+    if (arguments["sort"]) {
+        options.sort(arguments["sort"].get_document().value);
+    }
+    if (arguments["upsert"]) {
+        options.upsert(arguments["upsert"].get_bool().value);
+    }
+
+    stdx::optional<document::value> document;
+    switch (arguments["replacement"].type()) {
```

Comment:
Could use an `if` here; unsure what's idiomatic in cpp.

Reply:
The [ISO C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-switch-if), which Clyde has linked to in the past, recommends switch when there is an option. Compilers can also warn against unhandled cases.

For one case, I'd naturally do an `if/else`. But I don't have a strong preference.

In this particular case, since we only care that `replacement` is a document, I don't think an `if/else` or `switch` is required. `get_document` will throw an exception if `replacement` is not a document type.

Reply:
Gotcha, similar to Go then; the explanation they give about exhausting an enum whenever possible makes sense. Removed the `switch` here though as `get_document` should catch the relevant error.